spec/sneakers/integration_spec.rb

```diff
@@ -100,8 +85,352 @@ def start_worker(w)
+      cleanup_redis(redis)
+      prepare_sneakers(exchange: exchange_name)
+
+      worker.from_queue(queue_name, worker_opts)
+      @worker_pid = start_worker(worker)
+    end
+
+    after :each do
+      Process.kill('TERM', @worker_pid)
+
+      cleanup_rabbitmq(rabbitmq_client)
+      cleanup_redis(redis)
+    end
+
+    describe 'with defaults for maxretry handler' do
+      let(:worker) { MaxretryWorker }
+      let(:worker_opts) do
+        {
+          handler:         Sneakers::Handlers::Maxretry,
+          retry_max_times: 2,
+          retry_timeout:   100,
+          arguments: {
+            'x-dead-letter-exchange' => "#{queue_name}-retry"
+          }
+        }
+      end
+
+      it 'creates the required exchanges' do
+        expected_exchange_names = [
+          exchange_name,
+          "#{queue_name}-error",
+          "#{queue_name}-retry",
+          "#{queue_name}-retry-requeue"
+        ]
+        exchange_names = rabbitmq_client.list_exchanges.map(&:name)
+
+        expected_exchange_names.each do |expected_exchange_name|
+          assert_includes(exchange_names, expected_exchange_name)
+        end
+      end
+
+      it 'creates the required queues' do
+        expected_queue_names = [
+          queue_name,
+          "#{queue_name}-error",
+          "#{queue_name}-retry"
+        ]
+        queue_names = rabbitmq_client.list_queues.map(&:name)
+
+        expected_queue_names.each do |expected_queue_name|
+          assert_includes(queue_names, expected_queue_name)
+        end
+      end
+
+      describe 'when worker allways fails' do
+        let(:worker) { FailingMaxretryWorker }
+
+        before do
+          Sneakers::Publisher.new(
+            exchange: exchange_name
+          ).publish(
+            'foo',
+            routing_key: queue_name
+          )
+        end
+
+        it 'has a message in the error queue' do
+          # wait for failing message
+          sleep REQUEUING_WAIT_TIME
+
+          message = get_message_from_queue("#{queue_name}-error")
+
+          refute_nil(message.first)
+        end
+
+        it 'has been retried twice' do
+          # wait for failing message
+          sleep REQUEUING_WAIT_TIME
+
+          message = get_message_from_queue("#{queue_name}-error")
+
+          # Hmmmmm... should be 2
+          assert_equal(JSON.load(message[2])['num_attempts'], 3)
+        end
+      end
+
+      describe 'when worker fails once' do
+        let(:worker) { RetryOnceWorker }
+
+        before do
+          Sneakers::Publisher.new(
+            exchange: exchange_name
+          ).publish(
+            'foo',
+            routing_key: queue_name
+          )
+        end
+
+        it 'has no message in the error queue' do
+          # wait for failing message
+          sleep REQUEUING_WAIT_TIME
+
+          message = get_message_from_queue("#{queue_name}-error")
+
+          assert_nil(message.first)
+        end
+
+        it 'consumes the requeued message' do
```

Comment:
I think it would be good to ensure in this test that the retried message has the same routing key that it originally did. My initial concern was that by always providing a routing key on the dead letter exchange, the original routing key will be replaced when it is retried. I think for the case where you want to override each of these values, that is ok, but for the nominal case, you want to preserve the original message as much as possible.